### DISCUSSION TITLE
Using coldtype in a container

### DISCUSSION THREAD

**Discussion by vishalchalana** [coldtype/coldtype#84] [Q&A]
Hi,

I just discovered coldtype and exactly what I was looking for - love it! However, my use case is that I want to export the animation (as image sequence or gif or video) but I need to do it programmatically inside a Docker container. Is that possible?

Thanks,
Vish

**Comment by stenson** [comment]
Great to hear! I have to admit I don't know too much about docker, but if you want to render an animation from the command-line (without opening the coldtype viewer), you can do something like `coldtype filename.py -a` (-a for `render-all`). Is that close to what you had in mind?

**Reply by vishalchalana** [reply]
Thanks for your quick response @stenson. I did not know that this can be done through the command line also - that will surely help. But just to understand, what is the 'coldtype' command-line action actually doing? Is that something I can directly code?

One last question - is coldtype supported on Linux? 

I will try it on a container and let you know how it goes.

Thanks again!

**Reply by stenson** [reply]
The `coldtype` command-line program is the main way of interacting with coldtype — it encapsulates the "rendering" part of Coldtype. In standard use, the command-line program launches a viewing window for interactive programming (via file-change monitoring), but adding that `-a` instructs the rendering program to not open the viewing window and instead just render the animation frame-by-frame to disk. (A little bit more about that here: https://coldtype.goodhertz.com/tutorials/overview.html)

That said, it should be possible to do the same thing without the command-line, i.e. a bit of code like this:

```python
from coldtype.renderer import Renderer

_, parser = Renderer.Argparser()
parsed = parser.parse_args(["examples/animations/simplevarfont.py", "-a"])
Renderer(parsed).main()
```

_But_ that doesn't work quite yet, because the `Renderer` isn't smart enough to distinguish between an already-parsed `args` object and one that still needs to be parsed. I should be able to get that change into the next release of Coldtype, hopefully next week if all goes well with some current feature work I'm doing.

As for Linux, it should work as far as I know. I've used Coldtype quite a bit with the Google Colab notebook system, which is all linux-based, and I've also used it locally on a Chromebook on the sandboxed version of Linux provided there. Of course, some of the GUI parts of Coldtype might not function as well on Linux as on Mac, though that's only because I don't have a Linux machine to test on.

**Reply by vishalchalana** [reply]
Hi Rob,

I spent a lot of time working with coldtype in the last few days. I was able to implement a bunch of text animations that I wanted - _again, I am amazed how easy these were to implement using coldtype (thanks to you!)_. I integrated it with my service (video with animated text overlay) and was able to make it work on my local machine. I then deployed it in a docker container and was able to successfully do it by installing a few additional components. Here's where I am at.

- Because of the command-line interface only, it becomes really hard to 'parameterize' things. For example, in my use case, I have to pass animation types, colors, fonts, etc. Currently, the only way to do this was to pass these as command-line arguments which I parse and use. I did make it work but not an ideal implementation, as you can imagine.

- More importantly, I was unable to make it work on a container and I believe that the reason for this is that coldtype uses OpenGL and it seems there are limitations with Headless OpenGL. Here's the error I get.
`/usr/local/lib/python3.9/site-packages/glfw/__init__.py:906: GLFWError: (65544) b'X11: The DISPLAY environment variable is missing'`
It seems that GLFW does not support headless OpenGL at all but I am not sure about this. I tried multiple things but could not quite come up with a solution.

Do you think there is a workaround? Please let me know your thoughts.

Thanks,
Vish

**Reply by stenson** [reply]
Great to hear it's useful! Sorry for the delay in getting back to you — both issues definitely make sense. I think for the first one, I might do something like keep the state in a parallel json file, and then only pass minimal state through the command-line interface, if that makes sense, i.e. something like:

`data.json`

```json
{
    "0": "A",
    "1": "B",
    "2": "C"
}
```

`readjson.py`

```python
from coldtype import *
import json

key = __inputs__[0]
j = json.loads(__sibling__("data.json").read_text())
txt = j[key]

@animation()
def scratch(f):
    return (StSt(f"{txt}{f.i}", Font.RecursiveMono(), 300)
        .align(f.a.r, th=0))
```

And then call that with just a key value like this:

```sh
$ coldtype scratch/readjson.py 0
```

Still kind of a workaround, although the Coldtype architecture is currently much more oriented toward interactive programming than batch processing. Definitely something to think about for the future, though!

As for the GL/glfw issue, I think if you pass in the `-cpu` option to the command line, i.e. `coldtype filename.py -a -cpu`, then it should skip attempting to use OpenGL/GPU completely, and just render on the CPU. At least that's the intention — I don't use that option too frequently, but hopefully still works!

**Reply by vishalchalana** [reply]
Thanks, Rob. I am traveling so could not try these. I finally worked on it today and am happy to report that I was able to make it work in the container. It works perfectly! Thank you.

I have one more question. Is there more detailed documentation of coldtype somewhere? I have gone through https://coldtype.goodhertz.com/ and also your YouTube videos. I still don't clearly understand all the classes and methods.

I am currently stuck at one point. My timeline is, say 60 but I want to animate only for 30 frames. So, after 30 frames, I want a static image - the final frame of the animation. How to achieve this? Is it a parameter in `f.e` or something else?

**Reply by stenson** [reply]
Glad it works! I wish I could say there was more documentation at this point, but for the moment as the API isn't totally finalized, I've been hesitant to really sit down and write out the documentation, given that it gets out of date fairly quickly. I do intend to write out some proper documentation, but closer to the "1.0" release coldtype. That said, there are lot of examples in the repo, if you haven't checked those out: https://github.com/goodhertz/coldtype/tree/main/examples (most of them are in the `animations` folder)

As for you're specific question about the timing, I think something like this would work:

```python
from coldtype import *

@animation(timeline=60)
def scratch(f):
    t = Timeable(0, 30).at(f.i)

    return (StSt("TYPE"
        , Font.ColdtypeObviously()
        , 300
        , wdth=t.e("eeio"))
        .f(hsl(0.6))
        .align(f.a.r))
```

This uses a new function (`Timeable.at`) from the `0.8.0` release (so you might need to update to 0.8.0 or the latest release, 0.8.1), but should do the trick!

The idea here is to replace the `f` argument with your own `Easeable` object; the `at` function yields an `Easeable` object (the same type as the `f` argument) from any given `Timeable`, so you can use that as a drop-in replacement for the `f` argument. Yikes that's pretty complicated sounding when written out but hopefully makes some sense — basically just using a 30-frame timed element instead of the 60-frame timed element created automatically by the @animation decorator.

**Reply by vishalchalana** [reply]
That worked perfectly Rob. Thanks a lot!

Have another couple of questions now.
1. I am looking to give a 'fade-in' and 'fade-out' effect. I am able to do this by 'easing' alpha in `.f()`. However, I am wondering if it's possible to do it on `DATPens` directly keeping all `DATPen`s intact. Hope I am making sense here.
2. I am passing a whole text line to `StSt` but I want some words in different colors. Is that possible or do I have to draw it word by word? The problem with this approach is that I have to offset each word and then apply the same style to each word etc. It's doable but I am hoping for an easier way out.

Once again, thanks a lot for all your help.

**Reply by stenson** [reply]
Great! Very interesting questions again — there is supposed to be a way to do exactly what you're describe in (1), via the `.a(` method on DATPens, but it appears to be broken in 0.8.1 — will look into why that's not working (the idea is you could call `.a(0.25)` on the top-level of the hierarchy and it would cascade down throughout the whole thing). It does seem to work when you call it directly on sub-elements, so something like this code snippet might be pretty close to what you have in mind (if I'm understanding correctly):

```python
from coldtype import *
from coldtype.text.richtext import RichText

txt2 = """
Hello, [a]
WORD
"""

@animation((1080, 540), timeline=30)
def scratch(f):
    def dimmer(p, pos, _):
        if pos == 0:
            p.a(f.e("eeio"))

    return (RichText(
        f.a.r, txt2, {
            "a": Style(Font.RecursiveMono(), 200
                , ro=1
                , fill=hsl(0.9, a=0.25)
                , stroke=hsl(0.9, 1)
                , strokeWidth=2),
            "default": Style(Font.MutatorSans(), 150
                , fill=bw(0)
                , wght=1
                , wdth=1)})
        .xalign(f.a.r)
        .lead(20)
        .walk(dimmer)
        .align(f.a.r))
```

Also demonstrates the `RichText` module, which is a partial answer to your second question. Unfortunately that module's even less documented than the already under-documented main modules, but there are a few examples in the examples directory, usually with "richtext" in the file name.

---

Another option for coloring distinct words (if you want to avoid going down the `RichText` rabbit-hole) would be to `.split` the glyphs returned from a `StSt`, like this:

```python
from coldtype import *

@animation((1080, 540))
def scratch(f):
    return (StSt("Hello World".upper(), Font.MutatorSans(), 100, wght=1)
        .align(f.a.r)
        .split(lambda p: p.glyphName == "space")
        .index(0, lambda p: p.f(hsl(0.3)))
        .index(-1, lambda p: p.f(hsl(0.7))))
```

![image](https://user-images.githubusercontent.com/89770/148309848-f92605d8-08aa-4c47-8e77-b1f848087e49.png)

Not incredibly elegant, but does let you address and style individual words with the `.index` calls.

---

One other option, maybe more akin to what you’re already doing, is to "manually" compose StSt results with `.distribute`/`.track` for horizontal composition, or `.stack`/`.lead` for vertical compositions:

```python
from coldtype import *

@animation((1080, 540))
def scratch1(f):
    word1 = StSt("ONE", Font.MutatorSans(), 100).f(hsl(0.7))
    word2 = StSt("TWO", Font.MutatorSans(), 200).f(hsl(0.5))

    return (PS([word1, word2])
        .distribute()
        .track(10)
        .align(f.a.r))

@animation((1080, 540))
def scratch2(f):
    word1 = StSt("ONE", Font.MutatorSans(), 100).f(hsl(0.7))
    word2 = StSt("TWO", Font.MutatorSans(), 200).f(hsl(0.5))

    return (PS([word1, word2])
        .stack()
        .lead(10)
        .align(f.a.r))
```

![image](https://user-images.githubusercontent.com/89770/148309819-f6130566-61e9-4679-b0c1-e70075218d92.png)

**Reply by Mark2Mark** [reply]
That’s great! Is there a way to do something like this with multiline text? I tried a few-words-sting with newline chars `\n` instead of the space chars, and then the inspection with `.map(lambda p: print(p))` doesn’t print the `<®:P:RecordingPen(13mvs) {frame=Rect(2682.50,503.00,70.10,74.00),glyphName=H}>` for each glyph anymore, but instead only two items (one for each split at newline) like `<®:P:/4...> ...`. So how could it be done in a similar fashion?
I’d like to have a few words, each on a new line, and address them individually like in your one-line-examples here. Thanks so much!

**Reply by stenson** [reply]
Hmm, not totally sure I follow. Adding a newline increases the "depth" of the hierarchy returned. You can "flatten" the depth of the hierarchy with a `.collapse(deblank=False)` call (the `deblank` is to preserve the spaces rather than get rid of them), although that would get rid of the newline in the result, so you wouldn't be able to use that as data.

Is this kind of what you want? Rather than call the split on the top-level of the result, you map it over each line, meaning the resulting structure is something like (to use an xml-ish approximation):

```xml
<line>
  <slug>
    <glyph>A</glyph><glyph>B</glyph>
  </slug>
  <slug>
    <glyph>C</glyph><glyph>D</glyph>
  </slug>
</line>
<line>
  <slug>
    <glyph>E</glyph><glyph>F</glyph>
  </slug>
  <slug>
    <glyph>G</glyph><glyph>H</glyph>
  </slug>
</line>
```

```python
@animation((1080, 540))
def scratch(f):
    return (StSt("AB CD\nEF GH".upper(), Font.MutatorSans(), 200, wght=1)
        .xalign(f.a.r)
        .lead(30)
        .align(f.a.r)
        .map(lambda p: p.split(lambda p: p.glyphName == "space"))
        #.print()
        .index([0, 0], lambda p: p.f(hsl(0.3)))
        .index([0, 1], lambda p: p.f(hsl(0.04)))
        .index([1, 1], lambda p: p.f(hsl(0.7))))
```

<img width="652" alt="image" src="https://user-images.githubusercontent.com/89770/215143595-b6fc8aaf-84e3-4560-8a89-742acd7010aa.png">

**Reply by Mark2Mark** [reply]
Oh, that’s exactly what I had in mind. Thank you so much! The mapping via lambda to split things was the missing piece. This is awesome and really clean  🙌
